tools/grid/test/grid-memory.test.ts

```diff
@@ -0,0 +1,72 @@
+/*
+Copyright 2023 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+
+import { expect, fixture, nextFrame } from '@open-wc/testing';
+import { html, render } from '@spectrum-web-components/base';
+import { Default } from '../stories/grid.stories';
+
+async function usedHeapMB(): Promise<number> {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore
+    const memorySample = performance.measureUserAgentSpecificMemory();
+    return (await memorySample).bytes / (1024 * 1024);
+}
+
+describe('Grid memory usage', () => {
+    it('releases references on disconnect', async function () {
+        if (!window.gc || !('measureUserAgentSpecificMemory' in performance))
+            this.skip();
+
+        this.timeout(10000);
+
+        const iterations = 50;
+        let active = false;
+
+        const el = await fixture<HTMLElement>(
+            html`
+                <div></div>
+            `
+        );
+
+        async function toggle(
+            forced: boolean | undefined = undefined
+        ): Promise<void> {
+            active = forced != null ? forced : !active;
+            render(active ? Default() : html``, el);
+            await nextFrame();
+            await nextFrame();
+        }
+
+        // "shake things out" to get a good first reading
+        for (let i = 0; i < 5; i++) {
+            await toggle();
+        }
+        await toggle(false);
+        const beforeMB = await usedHeapMB();
+
+        for (let i = 0; i < iterations; i++) {
+            await toggle();
+        }
+        await toggle(false);
+        const afterMB = await usedHeapMB();
+
+        /**
+         * An actually leak here shapes up to be more than 10MB per test,
+         * we could be more linient later, if needed, but the test currently
+         * shows less heap after the test cycle.
+         */
+        expect(
+            afterMB - beforeMB,
+            `before: ${beforeMB}, after: ${afterMB}`
+        ).to.be.lt(0);
```

Comment:
You can also potentially use a WeakRef to the Grid element and wait a certain amount of time/ trigger garbage collection and expect the deref() to return undefined. This might be more deterministic than measuring heap memory.

Reply:
Opted for the more reliable, but slightly more cumbersome to surface, `[performance.measureUserAgentSpecificMemory()](https://chromestatus.com/feature/5685965186138112)` for now. Showing even more positive results from this change when measuring with this API.

I know the team has been looking at tooling with WeakMaps, I wonder how you're actually holding things in that way as it feels like the need for the key would inherently keep a reference and prevent the memory from being released. You wouldn't happen to have any examples to share?

Reply:
We're mainly using [WeakRef](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef#instance_methods) different from [WeakMap](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakMap). The basic idea is to construct a new WeakRef passing in the DOM node you're interested in tracking, trigger garbage collection, and then test whether or not calling WeakRef.deref() on your reference returns undefined. If so, you can. assume the node was garbage collected. It not [fool-proof](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/WeakRef#avoid_where_possible) but it has been working quite nicely for us.

Here is the original dom inspection snippets that Kin created: https://git.corp.adobe.com/jblas/debug-utils/tree/main/snippets#object-watcher

Reply:
This is cool, I didn't realize that WeakRef was different than WeakMap! Definitely a useful pattern for the future, particularly as a comparison for test speed for tightening the feedback cycle. 

Digging into `measureUserAgentSpecificMemory()` will also give us some other useful tidbits later on, for an expanded perf testing suite, so getting started now seems useful.